### PR TITLE
Add a new CUDA Next library and a first entry in it with hierarchy_dimensions type template

### PR DESCRIPTION
**Pull request by pciolkosz**
This change introduces a new CUDA Next library that aims to be a collection of experimental features. It would allow CUDA users to test and provide feedback on those features before they end up in other libraries in a more stable manner.
This change adds license file and some minimal cmake bits that needs to be extended in future changes.

The first entry in the CUDA Next library added in this pull request is hierarchy_dimensions type template that allows to describe a hierarchy of CUDA threads with a mix of static and dynamic information. It can be used to pass into kernels and then calculate aggregates at compile time (like count threads in each CUDA block to create a statically sized array). It can also be used to make libraries aware of the shape of currently running grid and optimize some thread id calculation with compile time values.

hierarchy_dimensions type template is basically a tuple of level_dimensions entries that describe levels. Each level consist of two things, first one is a type to describe what that level it is, for example block_level, cluster_level, etc. Second one is cuda::std::extents object to describe dimensions of that level with both static and dynamic values, the same way it describes cuda::std::mdspan objects.
This is an initial implementation, there is a number of TODOs, names are not final and the interface can change depending on the feedback received.

This type is is also a building block for other libraries and features that are in the pipeline.

NVBUG 4541889

### CODE REVIEW

File: cuda_next/CMakeLists.txt
Comment (miscco) [inline cuda_next/CMakeLists.txt:1]: That seems strange, we should not build that by default but only if requested
```suggestion
if (CCCL_ENABLE_CUDA_NEXT)
```

File: cuda_next/README.md
Comment (miscco) [inline cuda_next/README.md:7]:
```suggestion
**All APIs available in CUDA Next are not considered stable and can change without a notice.** They can also be deprecated or removed on a much faster cadence than in other CCCL libraries. 
```

File: cuda_next/README.md
Comment (miscco) [inline cuda_next/README.md:9]:
```suggestion
Features are exposed here for the CUDA C++ community to experiment with and provide feedback on how to shape it to best fit their use cases.
```

File: cuda_next/README.md
Comment (miscco) [inline cuda_next/README.md:10]:
```suggestion
Once we become confident a feature is ready and would be a great permanent addition in CCCL, it will become a part of some other CCCL library with a stable API.
```

File: cuda_next/include/detail/dimensions.hpp
Comment (miscco) [inline cuda_next/include/detail/dimensions.hpp:7]:
```suggestion
// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
```

I thoroughly hate dates in licenses

File: cuda_next/include/detail/dimensions.hpp
Comment (miscco) [inline cuda_next/include/detail/dimensions.hpp:20]: Important:
* We should be clear whether we see this as a user driven testbed or whether we already apply strict coding rules from libcu++. This would e.g require reserved identifiers etc

File: cuda_next/include/detail/dimensions.hpp
Comment (miscco) [inline cuda_next/include/detail/dimensions.hpp:23]: I am a bit scared by that alias in the `cuda_next` namespace

File: cuda_next/README.md
Comment (miscco) [inline cuda_next/README.md:12]: This is missing the minimum requirements. Otherwise it would inherit the default support matrix from CCCL which is a bit much I believe

File: cuda_next/include/detail/dimensions.hpp
Comment (miscco) [inline cuda_next/include/detail/dimensions.hpp:30]: Conventionally the `_CCCL_HOST_DEVICE` comes first, then the explicit
```suggestion
  _CCCL_HOST_DEVICE explicit constexpr hierarchy_query_result(const Dims& dims)
```

File: cuda_next/include/detail/dimensions.hpp
Comment (miscco) [inline cuda_next/include/detail/dimensions.hpp:26]: I am missing some documentation here. 

For exampleI was suprised that this directly inherits 3 dimensions

File: cuda_next/include/detail/dimensions.hpp
Comment (miscco) [inline cuda_next/include/detail/dimensions.hpp:33]: Should this also require `rank <= 3` At least more dimensions seem to be superfluous

File: cuda_next/include/detail/dimensions.hpp
Comment (miscco) [inline cuda_next/include/detail/dimensions.hpp:63]: Those two functions feel like an implementation detail and less than something we want to expose

File: cuda_next/include/detail/hierarchy_dimensions.hpp
Comment (miscco) [inline cuda_next/include/detail/hierarchy_dimensions.hpp:17]: I would prefer if we could put `<nv/target>` at the top

File: cuda_next/include/detail/hierarchy_dimensions.hpp
Comment (miscco) [inline cuda_next/include/detail/hierarchy_dimensions.hpp:110]: This is evil, overloading `operator&` is almost always not what we want. I would need to see a good usecase for this that would not be served by one of the less dangerous operators

Comment (miscco) [inline cuda_next/include/detail/hierarchy_dimensions.hpp:110]: Is there a reason a lot of the functions below are missing `_CCCL_HOST_DEVICE`?

**Review by miscco — COMMENTED** [review]
Some cursory comments, I did not dive into the code itself yet.

Generally I am super excited to get this started, but we should discuss some core guidelines expectations going we have

File: cuda_next/CMakeLists.txt
Comment (robertmaynard) [inline cuda_next/CMakeLists.txt:12]: This todo needs to be done completed before we can merge, so that users of CCCL can use this library.

File: cuda_next/CMakeLists.txt
Comment (robertmaynard) [inline cuda_next/CMakeLists.txt:15]: Testing needs to be optional. Please add an `option` like CCCL / libcudacxx does to allow people to opt out of tests. That will make this easier to package.

File: cuda_next/CMakeLists.txt
Comment (pciolkosz) [reply cuda_next/CMakeLists.txt]: I wanted tests compilation to work when running Cmake in CUDA_Next directory directly. Added CUDA_NEXT_TOPLEVEL_PROJECT similar to how other libraries do it instead.

File: cuda_next/README.md
Comment (pciolkosz) [reply cuda_next/README.md:12]: Added C++ standard requirement and minimal support matrix (what I used for development). I will expand it once I get the CI running for it

File: cuda_next/include/detail/dimensions.hpp
Comment (pciolkosz) [reply cuda_next/include/detail/dimensions.hpp:37]: I wanted hierarchy_query_result to be sort of a mix of dim3 and extents to make it easier to use for someone used to dim3, for example expressions like hierarchy.index().x or assigning to a dim3. It's probably OK if that is not a constexpr, it can still be accessed as constexpr with extents member functions.
Internally right now all extents objects always have 3 dimensions, because that is what block_dims and other functions creating levels return. I believe even if we end up supporting different number of dimensions we could still make this dim3-like behavior to work, either adding ones for the missing dimensions or collapsing extra ones into z

File: cuda_next/include/detail/dimensions.hpp
Comment (pciolkosz) [reply cuda_next/include/detail/dimensions.hpp:26]: See the combined comment below

File: cuda_next/include/detail/dimensions.hpp
Comment (pciolkosz) [reply cuda_next/include/detail/dimensions.hpp:50]: Yes, any dynamic extent will make merge result to also be dynamic. For example when X dimension of the block is 256 threads, but grid dimension X in terms of blocks is a dynamic value, then when expressing grid dimension X in terms of threads we can only say its some dynamic value

File: cuda_next/include/detail/dimensions.hpp
Comment (pciolkosz) [reply cuda_next/include/detail/dimensions.hpp:63]: They seems to be in the detail namespace unless I am mistaken

File: cuda_next/include/detail/dimensions.hpp
Comment (pciolkosz) [reply cuda_next/include/detail/dimensions.hpp]: Renamed to dimensions_index_type

File: cuda_next/include/detail/hierarchy_dimensions.hpp
Comment (pciolkosz) [reply cuda_next/include/detail/hierarchy_dimensions.hpp]: I think it should be consistent now

File: cuda_next/include/detail/hierarchy_dimensions.hpp
Comment (pciolkosz) [reply cuda_next/include/detail/hierarchy_dimensions.hpp]: I used & because it seemed ok semantically. Would operator* work better? I am not attached to any operator so if you have some suggestion I would be glad to hear it. I'm not even sure if we want the operator at all, but wanted to add it as optional and easier in-line hierarchy creation

File: cuda_next/include/detail/hierarchy_dimensions.hpp
Comment (pciolkosz) [reply cuda_next/include/detail/hierarchy_dimensions.hpp:381]: I wanted it here only for template argument deduction, similar how count or index queries take unit and level objects, but never use them

File: cuda_next/CMakeLists.txt
Comment (pciolkosz) [reply cuda_next/CMakeLists.txt]: Added CUDA_NEXT_ENABLE_TESTING option

File: cuda_next/CMakeLists.txt
Comment (pciolkosz) [reply cuda_next/CMakeLists.txt:12]: Added target_compile_features(CUDA_Next INTERFACE cxx_std_17) since we target C++17 with this library

File: cuda_next/CMakeLists.txt
Comment (pciolkosz) [reply cuda_next/CMakeLists.txt]: I added some minimal config files in lib/cmake/cuda_next to make it work. There is much more content in these files in other libraries, but I am not sure how much of that we need, at least initially

Comment (robertmaynard) [reply cuda_next/CMakeLists.txt]: Here are the rest of the changes you need so that consumers of `CCCL` can use CUDA_Next

```
diff --git a/cuda_next/lib/cmake/cuda_next/cuda_next-config.cmake b/cuda_next/lib/cmake/cuda_next/cuda_next-config.cmake
index b0bdb4436..2be26566c 100644
--- a/cuda_next/lib/cmake/cuda_next/cuda_next-config.cmake
+++ b/cuda_next/lib/cmake/cuda_next/cuda_next-config.cmake
@@ -2,8 +2,20 @@ if (TARGET CUDA_Next)
   return()
 endif()
 
+set(CUDA_NEXT_VERSION_MAJOR 0)
+set(CUDA_NEXT_VERSION_MINOR 1)
+set(CUDA_NEXT_VERSION_PATCH 0)
+set(CUDA_NEXT_VERSION_TWEAK 0)
+
+set(CUDA_NEXT_VERSION "${CUDA_NEXT_VERSION_MAJOR}.${CUDA_NEXT_VERSION_MINOR}.${CUDA_NEXT_VERSION_PATCH}.${CUDA_NEXT_VERSION_TWEAK}")
+
 # We don't ship with CTK, so no need for the ugly name (yet)
 add_library(CUDA_Next INTERFACE)
 target_compile_features(CUDA_Next INTERFACE cxx_std_17)
 
 target_include_directories(CUDA_Next INTERFACE ${CMAKE_CURRENT_LIST_DIR}/../../../include)
+
+if (NOT CUDA_Next_CONFIG)
+  set(CUDA_Next_CONFIG "${CMAKE_CURRENT_LIST_FILE}")
+endif()
+find_package_handle_standard_args(CUDA_Next CONFIG_MODE)
diff --git a/lib/cmake/cccl/cccl-config.cmake b/lib/cmake/cccl/cccl-config.cmake
index 882c67938..6c41159ae 100644
--- a/lib/cmake/cccl/cccl-config.cmake
+++ b/lib/cmake/cccl/cccl-config.cmake
@@ -101,7 +101,10 @@ foreach(component IN LISTS components)
         "${cccl_cmake_dir}/../../../cuda_next/lib/cmake/" # Source layout (GitHub)
                                                           # No install layout for now
     )
-    target_link_libraries(CCCL::CCCL INTERFACE CUDA_Next)
+    if (TARGET CUDA_Next AND NOT TARGET CCCL::CUDA_Next)
+      add_library(CCCL::CUDA_Next ALIAS CUDA_Next)
+      target_link_libraries(CCCL::CCCL INTERFACE CUDA_Next)
+    endif()
   else()
     message(FATAL_ERROR "Invalid CCCL component requested: '${component}'")
   endif()
   ```

**Comment by alliepiper** [comment]
Some assumptions I'm making as I update the CMake. Let me know if anything here is incorrect (@jrhemstad for vis):

1. The (case-sensitive) project name should be CudaNext:
  a. CMake targets will be `CudaNext::CudaNext`/`CCCL::CudaNext`
  b. CMake-generated variables will be `CudaNext_SOURCE_DIR`, `CudaNext_BINARY_DIR`, etc
  c. User-options will be `CudaNext_ENABLE_TESTING`, etc
2. CudaNext versioning will follow CCCL versioning.
3. Potential/allowed dependencies:
  a. CudaNext may depend on libcudacxx. Libcudacxx will never depend on CudaNext.
  b. Thrust/CUB may depend on CudaNext. CudaNext will never depend on Thrust/CUB.
  
1b and 1c are not consistent for our other projects for legacy reasons, so I want to make sure we stay consistent with this one.

**Comment by pciolkosz** [comment]
@alliepiper Regarding 1. I think it's a good name, I started with CUDA_Next, but CudaNext seems better.
For 2. I started with version 0.1, but we can just align with the CCCL versioning, probably easier this way.
Regarding 3. we don't plan to ship CUDA Next in the CTK right now, in general nothing should depend on CUDA Next, but CUDA Next can depend on any other library. For not it depends only on libcudacxx.

File: cuda_next/README.md
Comment (ahendriksen) [inline cuda_next/README.md:3]:
```suggestion
Some of them are still actively designed or developed and their API is evolving.
```

File: cuda_next/CMakeLists.txt
Comment (alliepiper) [reply cuda_next/CMakeLists.txt]: @robertmaynard is there a reason to define the version info in the config file, rather than relying on the config-version file?

Comment (alliepiper) [reply cuda_next/CMakeLists.txt]: Also @robertmaynard: Do you know if rapids-cmake will need us to add the header search / custom install dir stuff for this library?

**Comment by alliepiper** [comment]
Another question: I'm guessing that we'll want to install these headers under a prefix, e.g. `#include <cuda/next/foo.hpp>` instead of just directly in `include/`. Or would we prefer `<cudanext/...>` or `<cuda_next/...>`?

**Comment by ahendriksen** [comment]
> I think `<cuda/next/...>` is nicest personally so I'll go with that for now, but lmk if there's any other preferences.

I agree that `cuda/next` is nicest, but it does run the risk of potentially overlapping with the existing headers in `<cuda/*>`. To avoid confusion and potential conflict, perhaps we should steer for the slightly less nice `<cuda_next/...>`? This would also make it more clear that this is a separate library from libcu++.

File: cuda_next/CMakeLists.txt
Comment (robertmaynard) [reply cuda_next/CMakeLists.txt]: > is there a reason to define the version info in the config file, rather than relying on the config-version file?
The scope of variables defined in the version file are local, and not exposed to the caller

> Do you know if rapids-cmake will need us to add the header search / custom install dir stuff for this library?

Not currently as CUDANext says it doesn't support being installed.

**Comment by alliepiper** [comment]
@pciolkosz I've noticed that the current tests have a couple of issues:

1. Missing libcudacxx dependency, so some headers aren't found at compile time. Manually hacking in an include path for the tests WAR this issue (Added `include_directories(/path/to/libcudacxx/include`) to `cuda_next/test/CMakeLists.txt`). I'll fix this more robustly in my cmake changes.
2. With the dependency fixed, tests fail at runtime when built with optimizations (`cmake -DCMAKE_BUILD_TYPE="Release" ...`) or in debug mode (`cmake -DCMAKE_BUILD_TYPE="Debug" ...`). They only seem to pass when built when `CMAKE_BUILD_TYPE` is unset. Note that we use `Release` builds in our CI.

I can WAR this for now by unsetting the build type for CudaNext, but this should be fixed before we merged, since it usually means there's some UB somewhere. The runtime failure is:

```
Basic
  Device execution
-------------------------------------------------------------------------------
/home/coder/cccl/cuda_next/test/catch2_helpers/testing_common.hpp:75
...............................................................................

/home/coder/cccl/cuda_next/test/catch2_helpers/testing_common.hpp:103: FAILED:
  REQUIRE( (cudaLaunchKernelEx(&config, lambda_launcher<Dims, Lambda>, dims, lambda)) == cudaSuccess )
with expansion:
  98 == 0
```

To save some time, CUDA error 98 is "invalid device function". Maybe it's related to the lambda being used? I have verified that `--extended-lambda` is being passed to nvcc.

File: cuda_next/include/detail/dimensions.hpp
Comment (miscco) [reply cuda_next/include/detail/dimensions.hpp:20]: We decided that this is going to be a "user-land" library, so no __ugly names etc

File: cuda_next/README.md
Comment (alliepiper) [reply cuda_next/README.md:12]: I've added the current CI matrix to this in my changes, we can adjust as this settles out.

File: cuda_next/CMakeLists.txt
Comment (alliepiper) [reply cuda_next/CMakeLists.txt]: I've added install rules, but all targets/install rules are opt-in and disabled by default. I've left the header search stuff out for now, we can patch it in if we decide we want it later.

File: cuda_next/CMakeLists.txt
Comment (alliepiper) [inline cuda_next/CMakeLists.txt:6]: Before merging, we need to make sure that tests pass without this WAR.

File: cuda_next/README.md
Comment (alliepiper) [inline cuda_next/README.md:18]: CI is currently testing 12.0 and 12.3.

File: cuda_next/README.md
Comment (alliepiper) [inline cuda_next/README.md:20]: We'll need to update this once we get the CI matrix finalized.

File: ci/matrix.yaml
Comment (alliepiper) [inline ci/matrix.yaml:90]: We can prune this down if we decide to tighten up the minimum host compiler versions, etc.

File: ci/matrix.yaml
Comment (alliepiper) [inline ci/matrix.yaml:7]: This line and those above it show some of the reasons I'm not a fan of the name "cuda next". It sounds like we're talking about CTK. Things like `cuda_prev_min` and `cuda_next_min` being completely unrelated make the infra around this project name confusing and unintuitive.

I'm growing more and more fond of `cccl::preview`. It' not overloaded with anything, and "preview" implies instability, rather than "next" which implies release-candidate.

**Review by alliepiper — CHANGES_REQUESTED** [review]
Added the changes for CMake and CI.

The build/test cycle can be executed locally by running the `ci/test_cuda_next.sh` script.

The issue with `CMAKE_BUILD_TYPE` needs to be resolved before merging.

The current CI matrix is very broad and the current results can be used to help figure out what compiler versions are worth supporting. Currently only some of the GCC builds and the intel compiler builds are passing. The errors need to be investigated and either fixed if real issues, or the matrix updated if the failures are due to old compiler bugs/quirks.

File: cuda_next/CMakeLists.txt
Comment (pciolkosz) [reply cuda_next/CMakeLists.txt:6]: I will take a look, thank you for looking into this

**Comment by alliepiper** [comment]
Updated to make necessary changes to work with new CI.

**Comment by alliepiper** [comment]
@pciolkosz These last pushes add devcontainers for all of the new linux environments and enables the standard CCCL warning flags. I patched up the trivial warning fixes from the new flags + new containers, can you take a look at the remaining CI failures?

**Comment by pciolkosz** [comment]
pre-commit.ci autofix

**Comment by jrhemstad** [comment]
Long name "CUDA Experimental"
Short name "cudax"

Namespace: `cuda::experimental::`
Include path: `<cuda/experimental>`

In samples:
```
namespace cudax = cuda::experimental;
```

**Comment by ericniebler** [comment]
I think it would make sense to separate the c++ in this pr (the hierarchy dimensions) from the infra, and make a separate pr for the library bits. There should also be docs.

**Comment by alliepiper** [comment]
I'm currently integrating this PR with the changes coming in #1654.

**Comment by alliepiper** [comment]
pre-commit.ci autofix

**Comment by alliepiper** [comment]
pre-commit.ci autofix

**Review by alliepiper — APPROVED** [review]

**Comment by alliepiper** [comment]
The infrastructure is complete and CI is ready but disabled -- once the code is ready, just add it back in and uncomment the `cudax` entries in `ci/matrix.yaml`. I'll merge this once it clears CI.

**Comment by alliepiper** [comment]
/ok to test

...

**Pull request merged**
